python/ray/dag/compiled_dag_node.py

```diff
@@ -407,49 +416,59 @@ def _read(self) -> bool:
         assert self._intermediate_buffer is None
         exit = False
         try:
-            input_data = self.input_reader.read()
-            self.set_intermediate_buffer(input_data)
+            future = self.input_reader.read()
+            self.set_intermediate_buffer(future)
         except RayChannelError:
             # Channel closed. Exit the loop.
             exit = True
         return exit
 
-    def _compute(self, class_handle) -> bool:
+    def _compute(
+        self,
+        overlap_gpu_communication: bool,
+        class_handle,
+    ) -> bool:
         """
         Retrieve the intermediate result from the READ operation and perform the
-        computation. Then, cache the new intermediate result. The caller must ensure
-        that the last operation executed is READ so that the function retrieves the
-        correct intermediate result.
+        computation. Then, cache the new intermediate result.
 
         Args:
+            overlap_gpu_communication: Whether to overlap GPU communication with
+                computation during DAG execution to improve performance.
             class_handle: An instance of the class to which the actor belongs. For
                 example, the type of `class_handle` is <class 'xxxx.Worker'> if the
                 actor belongs to the `class Worker` class.
-
         Returns:
             True if system error occurs and exit the loop; otherwise, False.
         """
-        input_data = self.reset_intermediate_buffer()
+        input_futures = self.reset_intermediate_buffer()
         method = getattr(class_handle, self.method_name)
         try:
-            _process_return_vals(input_data, return_single_output=False)
+            _process_return_vals(input_futures, return_single_output=False)
```

Comment:
@stephanie-wang Looks like we could not limit calling `future.wait()` only inside `reset_intermediate_buffer()`, `_process_return_vals()` also needs to call `future.wait()` because:
1) `_process_return_vals()` is not only called in the actor execution loop, but also in driver when calling adag.execute().
2) The read value is from a list of channels, therefore it is a list of `DAGOperationFuture`s, so we need to wait on each of the futures rather than a single one. (`ListFuture` may be useful if we want to wrap them).

Is this a concern for you?

Reply:
Why can't `_process_return_vals` take the resolved values (as before) instead of `DAGOperationFuture`?

Reply:
Done.